test/functional/not_fn.cpp

```diff
@@ -0,0 +1,352 @@
+
+  Distributed under the Boost Software License, Version 1.0.
+  (See accompanying file LICENSE.md or copy at http://boost.org/LICENSE_1_0.txt)
+*/
+
+#include <slb/functional.hpp>
+
+#include <functional>
+#include <type_traits>
+#include <utility>
+
+#include "../catch.hpp"
+
+#define CHECK_NESTED(...)                                                      \
+  do {                                                                         \
+    INFO(__FILE__ "(" << __LINE__ << "): " #__VA_ARGS__);                      \
+    check_##__VA_ARGS__;                                                       \
+  } while (false)
+
+// [func.not_fn], function template not_fn
+
+// template <class F> unspecified not_fn(F&& f);
+
+struct Boolean {
+  int ret;
+  Boolean(int val) : ret(val) {}
+
+  int operator!() const noexcept { return ret; }
+};
+
+struct BooleanObj {
+  int base;
+  BooleanObj(int val) : base(val) {}
+
+  int operator!() & noexcept { return base - 1; }
+  int operator!() && noexcept { return base - 2; }
+  int operator!() const & noexcept { return base - 3; }
+// gcc finds calls on const rvalues ambiguous up to version 4.8.
+#if !defined(__GNUC__) || (__GNUC__ > 4) ||                                    \
+    ((__GNUC__ == 4) && (__GNUC_MINOR__ > 8))
+  int operator!() const && noexcept { return base - 4; }
+#endif
+};
+
+// Account for P0012: "Make exception specifications be part of the type
+// system".
+static constexpr bool p0012 = !std::is_same<void(), void() noexcept>::value;
+
+struct C {
+  BooleanObj obj;
+  C(int val) : obj(val) {}
+
+  Boolean fun(int base) noexcept(p0012) { return {base + 0}; }
+  Boolean cfun(int base) const noexcept(p0012) { return {base + 1}; }
+  Boolean lfun(int base) & noexcept(p0012) { return {base + 2}; }
+  Boolean rfun(int base) && noexcept(p0012) { return {base + 3}; }
+  Boolean clfun(int base) const& noexcept(p0012) { return {base + 4}; }
+// gcc finds calls on const rvalues ambiguous up to version 4.8.
+#if !defined(__GNUC__) || (__GNUC__ > 4) ||                                    \
+    ((__GNUC__ == 4) && (__GNUC_MINOR__ > 8))
+  Boolean crfun(int base) const&& noexcept(p0012) { return {base + 5}; }
+#endif
+};
+
+struct D : C {
+  D(int val) : C(val) {}
+};
+
+template <typename T, bool IsNothrow = true>
+struct smart_ptr {
+  T* ptr;
+  smart_ptr(T* ptr) : ptr(ptr) {}
+  T& operator*() const noexcept(IsNothrow) { return *ptr; }
+};
+template <typename T>
+using smart_ptr_throws = smart_ptr<T, false>;
+
+std::true_type const nothrows{};
+std::false_type const throws{};
+std::integral_constant<bool, p0012> const p0012_nothrows{};
+
+template <typename R, bool IsNothrow, typename F, typename... Args>
+void check_not_fn(R&& r,
+                  std::integral_constant<bool, IsNothrow>,
+                  F&& nf,
+                  Args&&... args) {
+  CHECK(std::forward<F>(nf)(std::forward<Args>(args)...) == r);
+  CHECK(std::is_same<decltype(nf(std::forward<Args>(args)...)), R>::value);
```

Comment:
We should forward the `nf` here I think, although it doesn't seem like we have any test cases that break this currently.

Reply:
We should add one then.